includes/utils/class-amp-validation-utils.php

```diff
@@ -41,16 +83,39 @@ class AMP_Validation_Utils {
 	public static function init() {
 		add_action( 'rest_api_init', array( __CLASS__, 'amp_rest_validation' ) );
 		add_action( 'edit_form_top', array( __CLASS__, 'validate_content' ), 10, 2 );
+		add_action( 'wp', array( __CLASS__, 'callback_wrappers' ) );
+		add_filter( 'amp_content_sanitizers', array( __CLASS__, 'add_validation_callback' ) );
+		add_filter( 'init', array( __CLASS__, 'post_type' ) );
 	}
 
 	/**
-	 * Tracks when a sanitizer removes an node (element or attribute).
+	 * Tracks when a sanitizer removes a node (element or attribute).
 	 *
-	 * @param DOMNode $node The node which was removed.
+	 * @param DOMNode     $node   The node which was removed.
+	 * @param string|null $source The source of the removed element.
 	 * @return void
 	 */
-	public static function track_removed( $node ) {
+	public static function track_removed( $node, $source = null ) {
 		self::$removed_nodes[] = $node;
+		if ( isset( $source ) && ! in_array( $source, self::$plugins_removed_nodes, true ) ) {
+			self::$plugins_removed_nodes[] = $source;
+		}
+	}
+
+	/**
+	 * Gets the source of the asset.
+	 *
+	 * This attempts to find if it's from a theme or a plugin.
+	 * In that case, it also returns the name of the theme or plugin.
+	 *
+	 * @param string $asset The asset path to search for the source.
+	 * @return array $source The asset type and source.
+	 */
+	public static function get_source( $asset ) {
+		preg_match( ':wp-content/(themes|plugins|)/(.+?)/:s', $asset, $matches );
```

Comment:
Since `WP_CONTENT_DIR` can be customized, I think this shouldn't be explicitly looking for `wp-content`.

Reply:
Finally, both the themes dir and plugins dir can be customized via `WP_PLUGIN_DIR`. So this should be looking at it as well.

Reply:
Likewise, `WPMU_PLUGIN_DIR` for `mu-plugins`.

Reply:
Good points. [This commit](https://github.com/Automattic/amp-wp/commit/ea16d64c6284a9b68522e9d7f1aa2b031f94bbd2) looks for themes, plugins, and mu-plugins. There's probably room to improve this, though.